numpy/core/src/multiarray/arraytypes.c.src

```diff
@@ -755,66 +756,230 @@ VOID_getitem(void *input, void *vap)
-        for (i = 0; i < n; i++) {
-            PyArray_Descr *new;
-            npy_intp offset;
-            key = PyTuple_GET_ITEM(names, i);
-            tup = PyDict_GetItem(descr->fields, key);
-            if (_unpack_field(tup, &new, &offset) < 0) {
-                ((PyArrayObject_fields *)ap)->descr = descr;
+        else if (PyArray_IsScalar(op, Void)){
+            srcdescr = ((PyVoidScalarObject*)op)->descr;
+            if (!PyDataType_HASFIELDS(srcdescr)) {
+                PyErr_SetString(PyExc_TypeError, msg_nostruct);
                 return -1;
             }
-            /*
-             * TODO: temporarily modifying the array like this
-             *       is bad coding style, should be changed.
-             */
-            ((PyArrayObject_fields *)ap)->descr = new;
-            /* remember to update alignment flags */
-            if ((new->alignment > 1)
-                    && ((((npy_intp)(ip+offset)) % new->alignment) != 0)) {
-                PyArray_CLEARFLAGS(ap, NPY_ARRAY_ALIGNED);
+            srcdata = ((PyVoidScalarObject*)op)->obval;
+        }
+
+        names_size = PyTuple_GET_SIZE(descr->names);
+
+        /* if we found the src dtype above, simply copy and return */
+        if (srcdescr != NULL) {
+            int ret;
+
+            /* Fast path if dtypes are equal */
+            if(PyArray_EquivTypes(srcdescr, descr)){
+                for (i = 0; i < names_size; i++) {
+                    /* neither line will fail, in principle */
+                    if (_setupField(i, descr, dummy, &offset)) {
+                        goto field_fail;
+                    }
+                    PyArray_DESCR(dummy)->f->copyswap(ip + offset,
+                                             srcdata + offset, 0, dummy);
+                }
+                return 0;
             }
-            else {
-                PyArray_ENABLEFLAGS(ap, NPY_ARRAY_ALIGNED);
+
+            /* Slow path */
+            ret = PyArray_CastRawArrays(1, srcdata, ip, 0, 0,
+                                        srcdescr, descr, 0);
+            if (ret != NPY_SUCCEED){
+                return -1;
             }
-            res = new->f->setitem(PyTuple_GET_ITEM(op, i), ip+offset, ap);
-            ((PyArrayObject_fields *)ap)->flags = savedflags;
-            if (res < 0) {
-                break;
+            return 0;
+        }
+
+        /* Otherwise it is a tuple or scalar.
+         *
+         * Strategy: to avoid clobbered arrays upon failure, we use two
+         * passes: First, copy+cast using setitem from the tuple to a
+         * buffer with the same dtype as the dst array (which can fail),
+         * then copy from the buffer to the dst (which should not fail).
+         *
+         * Because setitem expects a real array (not just a dummy) we need
+         * to create a buffer array which we then clobber.
+         */
+        Py_INCREF(descr);
+        buffer = (PyArrayObject*)PyArray_NewFromDescr(&PyArray_Type,
```

Comment:
See #7757. If you plan to reuse `descr` after this call you need another incref and then decref before function exit, but after use.